data/transactions/logic/eval.go

```diff
@@ -158,8 +173,10 @@ func stackValueFromTealValue(tv *basics.TealValue) (sv stackValue, err error) {
 // newly-introduced transaction fields from breaking assumptions made by older
 // versions of TEAL. If one of the transactions in a group will execute a TEAL
 // program whose version predates a given field, that field must not be set
-// anywhere in the transaction group, or the group will be rejected.
-func ComputeMinTealVersion(group []transactions.SignedTxn) uint64 {
+// anywhere in the transaction group, or the group will be rejected. In
+// addition, inner app calls must not call teal from before inner app calls were
+// introduced.
+func ComputeMinTealVersion(group []transactions.SignedTxnWithAD, inner bool) uint64 {
```

Comment:
while this is probably ok for the first revision, should we require that all teal invocation on inner transactions should share the same teal version as the caller to ensure that any two calls to the same program would result in the same outputs ? I honestly can't think of an attack vector here, but it would clearly be more "simplified". ( on the flip side, it would require contract authors to revise their contracts ).

Reply:
I don't think we need to enforce that.  The attack I was concerned was only concerned about the boundary between allowing and disallowing inner txns.  I was concerned that there might exist old code that used poor sources of randomness, believing that only the block proposer could take advantage of it by deciding whether to put it in the block based on outcome.  Inner transactions change that game, because a txn can trying invoking the app that uses poor randomness and, if it does not like the outcome, abort after the inner call, cancelling it.

Reply:
The change here from `SignedTxn` to `SignedTxnWithAD` isn't really harmful on it's own, but it would require that every single transaction ( even a non-app txn ), would now need to have it's data structure copied.

Reply:
Stateful logic processing needs to take a copy no matter what because it holds onto the copies and stores ApplyData there as it evaluates. You are right that we could avoid a copy in logicsig processing if we are careful enough.  Let me know if I should look into it.  That seems quite fast compared to costs normally associated with signature checking though.

Reply:
I don't consider this to be a blocker, but I have no idea what the performance implication would be. I'm pretty confident that having another method `ComputeMinTealVersionForInnerTransactions` would completely eliminate these concerns. Note that this method could be "cheaper" : there is no need to check for applications/rekeying support since this call could not be invoked before the application support anyway.